Tests/iaas/flavor-naming/flavor-add-extra-specs.py

```diff
@@ -0,0 +1,345 @@
+set the SCS extra_specs.
+
+On most clouds, to add properties (extra_specs) to flavors, you need to have
+admin power; this program will otherwise report the failed settings.
+You can can use this for testing, better use =n|--no-change.
+Add -d|--debug for more verbose output.
+
+(c) Kurt Garloff <garloff@osb-alliance.com>, 6/2024
+SPDX-License-Identifier: CC-BY-SA-4.0
+"""
+
+import os
+import sys
+import getopt
+import openstack
+
+from flavor_names import parser_v2, parser_v1, SyntaxV1, SyntaxV2, CPUTYPE_KEY, DISKTYPE_KEY
+from flavor_names import Flavorname, Main, Disk, outname
+# outname, Attr, Main, Disk, Hype, HWVirt, CPUBrand, GPU, IB, Flavorname,
+# Inputter, lookup_user_input, prettyname, CompatLayer
+
+# globals
+DEBUG = False
+QUIET = False
+NOCHANGE = False
+
+
+def usage(out):
+    "Output usage information (help)"
+    print(__doc__, file=sys.stderr)
+    sys.exit(out)
+
+
+def min_max_check(real, claim, valnm, flvnm, extra):
+    """Check whether property valnm real is at least claim.
+       Prints ERROR is lower and returns False
+       Prints WARNING if higher (and returns True)
+       Returns True if no problem detected.
+       For floats, we allow for 1% tolerance in both directions.
+       """
+    # 1% tolerance for floats (RAM)
+    if claim is float:
+        chkval = real*1.01
+        chkval2 = real*0.99
+    else:
+        chkval = real
+        chkval2 = real
+    if chkval < claim:
+        print(f"ERROR: Flavor {flvnm} claims {claim} {valnm}{extra}, but only has {real}. Needs fixing.",
+              file=sys.stderr)
+        return False
+    if chkval2 > claim:
+        print(f"WARNING: Flavor {flvnm} claims {claim} {valnm}{extra}, but overdelivers with {real}.",
+              file=sys.stderr)
+    return True
+
+
+def check_std_props(flavor, flvnm, extra=""):
+    """Check consistency of openstack props with parsed SCS name specs
+    Return no of errors found."""
+    errors = 0
+    #  vcpus
+    if not min_max_check(flavor.vcpus, flvnm.cpuram.cpus, "CPUs", flavor.name, extra):
+        errors += 1
+    #  ram
+    if not min_max_check(flavor.ram, flvnm.cpuram.ram*1024, "MiB RAM", flavor.name, extra):
+        errors += 1
+    #  disk
+    disksz = 0
+    if flvnm.disk:
+        disksz = flvnm.disk.disksize
+    if not min_max_check(flavor.disk, disksz, "GiB Disk", flavor.name, extra):
+        errors += 1
+    return errors
+
+
+def generate_name_v2(flavor, cpu_type, disk0_type):
+    """Generate an SCS- v2 name for flavor,
+    using cpu_type (and disk0_type if needed).
+    Returns string."""
+    cpuram = Main()
+    cpuram.cpus = flavor.vcpus
+    cpuram.cputype = cpu_type
+    cpuram.ram = int((flavor.ram+12)/512)/2.0
+    if flavor.disk:
+        disk = Disk()
+        disk.disksize = flavor.disk
+        disk.disktype = disk0_type
+        flv = Flavorname(cpuram, disk)
+    else:
+        flv = Flavorname(cpuram)
+    return outname(flv)
+
+
+def check_name_extra(flavor, ver, match, flname):
```

Comment:
I have a hard time following the intricate logic of this function. Also, there have been considerable changes to scs-0103-v1. The extra_spec `scs:name-vN` no longer requires a name for version N of the standard on flavor naming; it merely requires a name that is valid for ANY version of the standard.

Reply:
Oh, and I think it's a very bad idea to destructively update the `flavor.extra_specs` here and then rely on some other party for actually performing an update on the server. We need to collect the changes in a separate `dict` which can then be used to compute the necessary API calls, or, preferably, the command lines that the operator can then execute.

Reply:
> Oh, and I think it's a very bad idea to destructively update the `flavor.extra_specs` here and then rely on some other party for actually performing an update on the server. We need to collect the changes in a separate `dict` which can then be used to compute the necessary API calls, or, preferably, the command lines that the operator can then execute.

If we decide we want to output `openstack flavor set ...` commands, then collecting the changes seems a must.
I have no particularly strong opinion on it, except that I do NOT want to force operators to do cut'n'paste.

I don't see why updating the `flavor.extra_specs` fields in-place would be bad. We could however do the API call here right away rather than returning the need for it. There may have been thoughts on more complicated logic that lead to the split and looking at the code, I'm glad that does not seem to be needed.

Reply:
The problem with updating the official dict in place is that we can not trace afterwards what is only local and what is on the server. Therefore I would collect the changes in a dedicated data structure. A dict (which I suggested above) could work, but in light of the current discussion, I propose to use the _Command pattern_, that is, to introduce an abstract class that represents a change of the flavor, and it provides methods (a) for building a command-line representation of this change, and (b) for executing the change on the openstack connection.